src/raft/recv_install_snapshot.c

```diff
@@ -8,6 +8,348 @@
+ *
+ * 3. The Follower calculates [page_checksum_t] for its local
+ * persistent state. When this process ends it puts itself into
+ * SIGNATURES_CALC_DONE state by calling follower_tick(..., NULL).
+ *
+ * 4. The Follower sends Signature() messages to the Leader, transfers
+ * the whole [page_checksum_t] and puts itself into
+ * SIGNATURES_PART_SENT state. When all signatures transferred the
+ * Leader, it sends Signature(done=true) message.
+ *
+ * 5. When the Leader received the first Signature() message it moves
+ * corresponding state machine into SIGNATURES_CALC_STARTED state and
+ * creates HT. The Leader puts incomming payloads of Signature()
+ * message into the HT. When the Leader received Signature(done=true)
+ * message it moves into SNAPSHOT_INSTALLATION_STARTED state.
+ *
+ * 6. In SNAPSHOT_INSTALLATION_STARTED state, the Leader starts
+ * iterating over the local persistent state, and calculates the
+ * checksum for each page the state has. Then, it tries to find the
+ * checksum it calculated in HT. Based on the result of this
+ * calculation, the Leader sends InstallShapshot(CP..) or
+ * InstallShapshot(MV..) to the Follower.
+ *
+ * Upon receving these messages the Follower moves into
+ * SNAPSHOT_CHUNCK_RECEIVED state. The Leader moves into
+ * SNAPSHOT_CHUNCK_SENT state after receiving first reply from the
+ * Follower.
+ *
+ * 7. When the iteration has finished the Leader sends
+ * InstallShapshot(..., done=true) message to the Follower. It moves
+ * the Follower back to NORMAL state and the state machine
+ * corresponding to the Follower on the Leader is moved to
+ * SNAPSHOT_DONE_SENT state.
+ *
+ * 8. The Leader sends AppendEntries() RPC to the Follower and
+ * restarts the algorithm from (1). The Leader's state machine is
+ * being moved to FOLLOWER_ONLINE state.
+ *
+ * =Failure model
+ *
+ * ==Unavailability of the Leader and Follower.
+ *
+ * To handle use-cases when any party of the communication becomes
+ * unavailable for a while without crash the following assumtions are
+ * made:
+ *
+ * - Signature() or InstallSnapshot(MV/CP) messages are idempotent and
+ *   can be applied to the persistent state many times resulting the
+ *   same transition.
+ *
+ * - Each message with data chuncks has an information about the
+ *   "chunk index". Chunk indexes come in monotonically increasing
+ *   order.
+ *
+ * - Each reply message acknowledges that the data received (or
+ *   ignored) by sending `result` field back to the counter part along
+ *   with last known chunk index as a confirmation that the receiver
+ *   "knows everything up to the given chunck index".
+ *
+ * - If a party notices that last known chunk index sent back to it
+ *   doesn't match it's own, the communication get's restarted from
+ *   the lowest known index.
+ *
+ * ==Crashes of the Leader and Follower.
+ *
+ * @TODO: We'd need to review other options here.
+ *
+ * The Follower maintains persistent bootcounter variable incremented
```

Comment:
Instead of persisting it, would it be enough to create a random number on start-up and communicate it to the leader? (assuming we need to communicate the bootcounter to the leader through some mechanism). When the random number changes that means a reboot has happened.